### PR TITLE
Point the user towards multiple_linear_regression if they use nd-y

### PR DESCRIPTION
**Pull request by MomoLangenstein**
Adds a helpful error message hint if someone tries to use a multidimensional `Y` with `linear_regression`

### CODE REVIEW

**Comment by Aggrathon** [comment]
The reasons for having different methods for (single) linear regression and multiple regression are:
a) Trying to be too helpful/smart can introduce hard to find bugs for the user.
b) PyTorch throws warnings about not being able to trace tensor->bool. This is a harmless warning in an assert, but creates a worse user experience.

**The second issue also affects this solution!** But it could maybe be solved by filtering the warning (see e.g. https://github.com/pytorch/pytorch/issues/19364). And if that works we might consider combining the two methods.

**Comment by MomoLangenstein** [comment]
> The reasons for having different methods for (single) linear regression and multiple regression are:
> a) Trying to be too helpful/smart can introduce hard to find bugs for the user.

I'd be more nuanced and say that being too smart by trying to guess user intent to "just make things work" is indeed error-prone. Therefore, I agree that having two methods is a good idea. However, one is the given default and unless you search through the source code, it's difficult to stumble across the other method, especially given the weird error pytorch throws in that case.

Hence, I think it is a good idea to catch such behaviour and give the user a helpful error message. Then, they don't first try to fix the slisemap source code, like I did, which is beneficial for creating a more convenient and less error-prone experience.

> b) PyTorch throws warnings about not being able to trace tensor->bool. This is a harmless warning in an assert, but creates a worse user experience.

Huh, I didn't see that during local testing. Maybe this would be fixed if the check is moved outside the loss function and into the initialiser (i.e. check if y is multidimensional and the local model is still `linear_regression`)?

**Comment by Aggrathon** [comment]
If you check the pytest logs (from the GitHub action) the warning appears.

 I think the correct place for the assert is in the loss function, but we could make it more readable with the following helper function:
```python
def _assert_no_trace(condition: Callable[[], bool], message: str):
    with warnings.catch_warnings():
        warnings.filterwarnings("ignore", category=torch.jit.TracerWarning)
        _assert(condition(), message)
```
Which would then be used like:
```python
_assert_no_trace(lambda: X.shape[1] == B.shape[1], "...")
```
Instead of:
```python
_assert(X.shape[1] == B.shape[1], "...")
```

**Comment by MomoLangenstein** [comment]
> If you check the pytest logs (from the GitHub action) the warning appears.
> 
> I think the correct place for the assert is in the loss function, but we could make it more readable with the following helper function:
> 
> ```python
> def _assert_no_trace(condition: Callable[[], bool], message: str):
>     with warnings.catch_warnings():
>         warnings.filterwarnings("ignore", category=torch.jit.TracerWarning)
>         _assert(condition(), message)
> ```
> 
> Which would then be used like:
> 
> ```python
> _assert_no_trace(lambda: X.shape[1] == B.shape[1], "...")
> ```
> 
> Instead of:
> 
> ```python
> _assert(X.shape[1] == B.shape[1], "...")
> ```

Fixed in 6885dc2782040c816c5ae7bed08af5c34d5b018e

**Comment by Aggrathon** [comment]
Looks good, just a minor nitpick: `"The B matrix has too many columns..."` is not actually what you are testing for (it could be too few as well).

**Comment by MomoLangenstein** [comment]
> Looks good, just a minor nitpick: `"The B matrix has too many columns..."` is not actually what you are testing for (it could be too few as well).

Thanks, good catch!
